### DISCUSSION TITLE
`brew info --json=v2 --all` times out on arm64/macos11.2.3-arm64

### DISCUSSION THREAD

**Discussion by haakonstorm** [Homebrew/discussions#1194] [Tap maintenance and Homebrew development] [answered]
<img width="798" alt="Screenshot of Firefox Developer Edition (04-04-2021, 22-19-48)" src="https://user-images.githubusercontent.com/193781/113520518-e46cd180-9593-11eb-8a16-c5a0ddc30b56.png">

## setup

```bash
λ uname -a; zsh --version; brew --version
Darwin smba13.local 20.3.0 Darwin Kernel Version 20.3.0: Thu Jan 21 00:06:51 PST 2021; root:xnu-7195.81.3~1/RELEASE_ARM64_T8101 arm64
zsh 5.8 (arm-apple-darwin20.2.0)
Homebrew 3.0.10-222-g77a53d6
Homebrew/homebrew-core (git revision f2f80f72d5; last commit 2021-04-04)
Homebrew/homebrew-cask (git revision 9169a5fc30; last commit 2021-04-04)
```

## brew config
```bash
λ brew config; brew doctor;

Homebrew/homebrew-core (git revision f2f80f72d5; last commit 2021-04-04)
Homebrew/homebrew-cask (git revision 9169a5fc30; last commit 2021-04-04)
HOMEBREW_VERSION: 3.0.10-222-g77a53d6
ORIGIN: https://github.com/Homebrew/brew
HEAD: 77a53d6730767bdf1289c50bf85b7a741d6fe083
Last commit: 12 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: f2f80f72d58bf8cfea313f21026c928b2e629c49
Core tap last commit: 9 hours ago
Core tap branch: master
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_BAT: set
HOMEBREW_CASK_OPTS: ["--appdir=/Applications", "--no-quarantine"]
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.qAYNCxfjHV/org.macosforge.xquartz:0
HOMEBREW_FORCE_BREWED_CURL: set
HOMEBREW_GITHUB_API_TOKEN: set
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.3 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit arm_firestorm_icestorm
Clang: 12.0 build 1200
Git: 2.31.1 => /opt/homebrew/bin/git
Curl: 7.76.0 => /opt/homebrew/opt/curl/bin/curl
macOS: 11.2.3-arm64
CLT: 12.4.0.0.1.1610135815
Xcode: 12.4
XQuartz: 2.7.11 => /opt/X11
Rosetta 2: false
Your system is ready to brew.
```

1. I'm trying to fetch a list of all formulae and casks in JSON format; as I've done previously, using `brew info --json=v2 --all`.
2. Same result whether Apple-bundled `curl` or homebrew-provided curl via `HOMEBREW_FORCE_BREWED_CURL` is used.
3. This appear to only happen on Big Sur. I tested on another Mac in another site, over ssh, and the list was fetched. This is a Catalina server in production.
4. Since this worked, I assumed the problem was local to my machine, so I created a new temporary admin account in macOS Big Sur, and tried again, but got same timeout.
5. Trying to find the source for the "tla.msr-inria.inria.fr" I `ripgrep`d my Taps directory:

```bash
λ rg 'inria'

(...)
homebrew/homebrew-cask-versions/Casks/tla-plus-toolbox-nightly.rb
6:base_url = "https://tla.msr-inria.inria.fr/tlatoolbox/ci/products/"
```

So now we know the `homebrew/homebrew-cask-versions/Casks/tla-plus-toolbox-nightly.rb` contains a link that 404s, and maintainers of this should be notified.

- But, should `brew` itself perhaps handle this error a bit more gracefully? Should I file a bug report?

- Also, is it necessary for `brew info --all` to actually perform all these live network requests? Maybe I'm shooting with too big of a cannon. I need to read more of the docs; but after running `brew update` shouldn't all the info needed to generate the `--json=v2` already be found locally?
 
- (In hindsight, since `--json=v1` worked, I should have started looking into casks / cask-versions earlier, but I wasn't sure whether `inria.fr` hosted all, or only parts, of what `brew` set out to find. Or whether `infria.fr` was part of a CDN on behalf of Microsoft/GitHub, and the issue only local to my part of the Net. (I'm in Norway, btw).)

I'm building a (shell) TUI of sorts which enables `fzf`-browsing of formulas, casks, npm, pypi, go releases and such with some `awesome-list` filtering/matching.

**Comment by SMillerDev** [comment, marked as answer]
> But, should brew itself perhaps handle this error a bit more gracefully? Should I file a bug report?

That seems reasonable

> Also, is it necessary for brew info --all to actually perform all these live network requests? Maybe I'm shooting with too big of a cannon. I need to read more of the docs; but after running brew update shouldn't all the info needed to generate the --json=v2 already be found locally?

Why not just use the API at https://formulae.brew.sh/ ?

**Comment by Bo98** [comment]
> is it necessary to actually perform all these live network requests

I believe this is limited to a select few casks which obtain their download URL dynamically. It's probably worth opening an issue about handling this better in `brew info`.

**Comment by haakonstorm** [comment]
Thanks a lot guys. I thought `brew info --all` contained additional metrics; like "popularity/number of fetches last 30 etc" as opposed to fetching the `formulae.json`. But they seem identical. At least the formulae. 
The output of most stuff in `brew` could benefit from `--json` as well, as in analytics:

<img width="1278" alt="Screenshot of iTerm2 (05-04-2021, 22-50-25)" src="https://user-images.githubusercontent.com/193781/113629117-38e08180-9666-11eb-94f1-f0100659a5db.png">

That this produce a table output and not either json output or an error, could also be considered an issue but mostly annoying, esp considering the API is there. I'm serving myself a proper plate of the docs tonight everyone, thanks again.
